plaidml/bridge/openvino/ops/roi_pooling.cpp

```diff
@@ -14,70 +14,109 @@ using namespace InferenceEngine;  // NOLINT[build/namespaces]
-  auto x_2 = coord[2];
-  auto y_2 = coord[3];
+edsl::Tensor crop_max_pooling(edsl::Tensor I, const std::vector<float>& coord, int64_t pooled_h, int64_t pooled_w,
+                              int height, int width) {
+  auto roi_w_start = coord[0];
+  auto roi_h_start = coord[1];
+  auto roi_w_end = coord[2];
+  auto roi_h_end = coord[3];
 
-  auto roi_width = std::max(x_2 - x_1, 1.0f);
-  auto roi_height = std::max(y_2 - y_1, 1.0f);
+  float roi_width = std::max(roi_w_end - roi_w_start + 1, 1.0f);
+  float roi_height = std::max(roi_h_end - roi_h_start + 1, 1.0f);
 
-  auto bin_size_h = roi_height / static_cast<float>(pooled_h);
-  auto bin_size_w = roi_width / static_cast<float>(pooled_w);
+  float bin_size_h = roi_height / pooled_h;
+  float bin_size_w = roi_width / pooled_w;
 
   edsl::Tensor pooled_tensor;
   for (auto i = 0; i < pooled_h; i++) {
     for (auto j = 0; j < pooled_w; j++) {
       // enlarge bin.
-      auto start_h = std::floor(x_1 + bin_size_h * i);
-      auto start_w = std::floor(y_1 + bin_size_w * j);
-      auto end_h = std::ceil(x_1 + bin_size_h * (i + 1));
-      auto end_w = std::ceil(y_1 + bin_size_w * (j + 1));
-
-      auto h_tensor =
-          edsl::index({edsl::TensorDim(static_cast<int64_t>(end_h - start_h))}, 0) + static_cast<int>(start_h);
-      auto w_tensor =
-          edsl::index({edsl::TensorDim(static_cast<int64_t>(end_w - start_w))}, 0) + static_cast<int>(start_w);
-
-      // crop bin from ROI.
-      auto gather_w = edsl::gather(I, w_tensor).axis(3).interpolationMode(edsl::InterpolationMode::LINEAR);
-      auto crop = static_cast<edsl::Tensor>(
-          edsl::gather(gather_w, h_tensor).axis(2).interpolationMode(edsl::InterpolationMode::LINEAR));
-
+      int h1 = roi_h_start + std::floor(bin_size_h * i);
+      int w1 = roi_w_start + std::floor(bin_size_w * j);
+      int h2 = roi_h_start + std::ceil(bin_size_h * (i + 1));
+      int w2 = roi_w_start + std::ceil(bin_size_w * (j + 1));
+
+      // check border.
+      auto start_h = std::min(std::max(h1, 0), height);
+      auto start_w = std::min(std::max(w1, 0), width);
+      auto end_h = std::min(std::max(h2, 0), height);
+      auto end_w = std::min(std::max(w2, 0), width);
+
+      // if start equal to end, we have to guarantee that index is at least one. and they are not over the border.
+      auto slice_w_index = end_w - start_w > 0 ? end_w - start_w : 1;
+      auto slice_h_index = end_h - start_h > 0 ? end_h - start_h : 1;
+      auto h_index = start_h >= height ? height - 1 : start_h;
+      auto w_index = start_w >= width ? width - 1 : start_w;
+      auto h_tensor = edsl::index({edsl::TensorDim(slice_h_index)}, 0) + h_index;
+      auto w_tensor = edsl::index({edsl::TensorDim(slice_w_index)}, 0) + w_index;
+      auto gather_w = edsl::gather(I, w_tensor).axis(3);
+      auto crop = static_cast<edsl::Tensor>(edsl::gather(gather_w, h_tensor).axis(2));
       // get max value from bin, then put it to pooled tensor.
```

Comment:
Please add new line here before the comment.

Reply:
done